### PR TITLE
ENH: Statespace: Add Kalman smoother

### PR DESCRIPTION
**Pull request by ChadFulton**
Adds pure python Kalman state and disturbance smoothers. Probably wouldn't have enough performance for use real use in estimation (e.g. EM or MCMC), but will be useful for people wanting to get smoothed output when parameters are estimated via MLE.

Also does not currently support smoothing in the case of missing data.

### CODE REVIEW

**Comment by ChadFulton** [comment]
Here's a notebook with a basic example using the smoothed output:

http://nbviewer.ipython.org/gist/ChadFulton/c3bc720c2fa31c564844/

**Comment by josef-pkt** [comment]
just a quick clarification to the description:

"pure python" means it doesn't use the cython extension?
It looks to me you still import it.

**Comment by ChadFulton** [comment]
To get the filtered output (or for parameter estimation, etc.), it uses the usual Cython extension.

It's just that the smoothing itself is done in pure python (see the `_kalman_smooth` function at the top of the kalman_smoother.py file).

**Comment by ChadFulton** [comment]
> Probably wouldn't have enough performance for use real use in estimation (e.g. EM or MCMC), but will be useful for people wanting to get smoothed output when parameters are estimated via MLE.

Whereas MLE estimation only calls the Kalman filter repeatedly, EM and MCMC call both the Kalman filter and smoother repeatedly. So while the Kalman filtering would still be fast, the Kalman smoothing would be slow here. Eventually I'll add a Cython Kalman smoother, but for now I thought something was better than nothing (and since we don't focus on EM or MCMC, the slower python implementation won't be noticeable).

**Comment by josef-pkt** [comment]
Ok, thanks for clarifying

brief look at the notebook

```
res_ll = mod_ll.fit()
res_ll = mod_ll.smooth()
```
looks strange because in python this looks like fit() would be thrown away, the first res_ll will be overwritten.

should this be
```
res_ll = mod_ll.fit()
res_ll = mod_ll.smooth(result=res_ll)
```
or is there a "state" of the model so it is equivalent to 

```
mod_ll.fit()
res_ll = mod_ll.smooth()
```

**Comment by ChadFulton** [comment]
Right now there is a "state" of the model, so that you're right, it's equivalent to:

```python
mod_ll.fit()
res_ll = mod_ll.smooth()
```

I know we'd like to move away from holding state like that. I plan (and I think we discussed before) to do that by separating the Statsmodels component (MLEModel, Model classes) from the Kalman filtering component (KalmanSmoother, KalmanFilter, Representation classes) and using delegation. Then the Statsmodels component won't hold state, but the Kalman filtering component will (it has to, since it holds the state space matrices which by definition hold state).

After that's done, I suppose that it will look something like your second example. Either

```python
res_ll = mod_ll.fit()
smoothed_ll = mod_ll.smooth(res_ll)
```

or:

```python
res_ll = mod_ll.fit()
smoothed_ll = mod_ll.smooth(res_ll.params)
```

**Comment by ChadFulton** [comment]
Test failure related to old Numpy version (same function failing as in #2431)

**Comment by ChadFulton** [comment]
Rebased so tests pass.

**Comment by ChadFulton** [comment]
Rebased after decoupling, etc. PRs.

**Comment by ChadFulton** [comment]
Rebased again.

**Comment by ChadFulton** [comment]
Rebased. I think this should be good to merge, the interface issues we discussed above have been fixed (I think) since the rebase after decoupling. See e.g. this notebook for the interface to get smoothed results: http://nbviewer.ipython.org/gist/ChadFulton/06557357b2be654b13a1

File: statsmodels/tsa/statespace/kalman_smoother.py [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+            (including state, disturbances, and all covariances).
+        results : class or object, optional
+            If a class, then that class is instantiated and returned with the
+            result of both filtering and smoothing.
+            If an object, then that object is updated with the smoothing data.
+            If None, then a SmootherResults object is returned with both
+            filtering and smoothing results.
+        prefix : string
+            The prefix of the datatype. Usually only used internally.
+        Returns
```
Comment (josef-pkt) [inline statsmodels/tsa/statespace/kalman_smoother.py]: blank line before

**Comment by josef-pkt** [comment]
similar here

looks good from a superficial review, example notebook looks nice.
But we will need feedback after playing with it for a while.

I started to read the Stata documentation for the models related to these PRs, but I would need to read again for a few days to "understand" this.

I'm fine with merging it as is.

File: statsmodels/tsa/statespace/kalman_smoother.py (hunk 1 above)
Comment (ChadFulton) [inline statsmodels/tsa/statespace/kalman_smoother.py]: Thanks.

**Comment by ChadFulton** [comment]
Rebased on master.

**Comment by josef-pkt** [comment]
travis the first py2.7 timed out, otherwise green

merging

**Pull request merged**